components/supervisor/pkg/terminal/terminal.go

```diff
@@ -146,44 +149,63 @@ func (m *Mux) doClose(alias string, gracePeriod time.Duration) error {
 	return nil
 }
 
-func (term *Term) gracefullyShutdownProcess(gracePeriod time.Duration) error {
+func (term *Term) gracefullyShutdownProcess(wg *sync.WaitGroup) error {
 	if term.Command.Process == nil {
-		// process is alrady gone
+		// process is already gone
 		return nil
 	}
-	if gracePeriod == 0 {
-		return term.shutdownProcessImmediately()
-	}
-
-	err := term.Command.Process.Signal(unix.SIGTERM)
-	if err != nil {
-		return err
-	}
-	schan := make(chan error, 1)
+	pid := term.Command.Process.Pid
+	wg.Add(1)
 	go func() {
-		_, err := term.Wait()
-		schan <- err
-	}()
-	select {
-	case err = <-schan:
-		if err == nil {
-			// process is gone now - we're good
-			return nil
+		defer wg.Done()
+		waitForChildProcessesToTerminate(pid)
+		err := unix.Kill(pid, unix.SIGTERM)
+		if err != nil {
+			return
 		}
-		log.WithError(err).Warn("unexpected terminal error")
-	case <-time.After(gracePeriod):
-	}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			_, err := term.Wait()
+			if err != nil {
+				log.WithError(err).Debug("error waiting for terminal")
+			}
+		}()
+	}()
 
-	// process did not exit in time. Let's kill.
-	return term.shutdownProcessImmediately()
+	return nil
 }
 
-func (term *Term) shutdownProcessImmediately() error {
-	err := term.Command.Process.Kill()
-	if err != nil && !strings.Contains(err.Error(), "os: process already finished") {
-		return err
+func waitForChildProcessesToTerminate(PID int) {
```

Comment:
The function doesn't do what it says on the tin. It sends sigkill rather than wait for the termination. Is that behaviour intentional?

Reply:
where does it send sigkill?